README.md
# Run locally

## Dependancies
```
pip install -r requirements.txt
```

## Start
```python
python.exe server.py
```

## UI
http://localhost:8000/

## Comment
First run may be slow because of the initialization of everything. 

There's no concurrency protection, click multiple times on Generate button at your own peril.

static/history
very attractive Beautiful nude 18 years old girl, shy smile , lying in bed, (((full body))) (((bukkake))):1.6, (((creampie))):1.9, (((she is having anal sex with long dildo))):1.9, (((she is having vaginal sex with long dildo))):1.9, (((dildo penetrating wet anus))):1.7, ((huge amount of dripping sperm inside pussy)):1.7, (((anus exposed))), (((bukkake))):1.6, ((huge amount of dripping sperm on legs and belly))1.7, virgin, spreading legs, legs up, perfect pussy, angelic woman, (acne:0.4),(perfect breast), (((hands behind head)))1.7, nsfw, 8k, GIRL,((( PERFECT FACE))), DETAILED PUPILS, ,FEMININE, (HIGH DETAIL), LONG HAIR, best quality, full body,high quality, realistic lighting, symmetrical natural eyes, fascinating makeup, long legs, pale skin, stunning lady, sensual mouth, natural eyes, beautiful realistic face, glamour photo, vogue style, RAW, high detail, sharp focus, smooth, aesthetic, extremely detailed, photo_\(ultra\), photorealistic, realistic, post-processing, max detail, roughness, real life, ultra realistic, photorealism, photography, 8k uhd, photography, perfect hands, nipples, french manicure, (big breasts, pink nipples:1.2)
3d rendering of Shrek in a superman costume, pixar
3d rendering of Shrek in a batman costume, pixar
DLSR photoshoot of Shrek in a batman costume
DLSR photoshoot of batman in a fastfood
DLSR photoshoot of Spiderman in a ironman costume
A drawing of cailloux
A child's drawing of a couple in front of a house
A children's drawing of a couple with a child
Landscape, masterpiece, intricate details
Painting of a wooden ship wreckage under water, ocean floor, masterpiece, intricate details
Painting of a wooden ship wreck under water, masterpiece, intricate details
Painting of submarine village under glass domes, fishes, masterpiece, intricate details
Painting of submarine village, kelp forest, glass dome, fishes, masterpiece, intricate details
Painting of submarine village, kelp forest, glass dome, masterpiece, intricate details
Painting of an elf clear stone single building, dark pine forest, cyberpunk, masterpiece, intricate details
Painting of an elf clear stone single building, dark pine forest, starcraft, masterpiece, intricate details
Painting of an elf clear stone single building, dark pine forest, warhammer, masterpiece, intricate details
Painting of an elf clear stone single building, dark pine forest, warcraft, masterpiece, intricate details
Painting of an elf clear stone single building, dark pine forest, magic the gathering, masterpiece, intricate details
Painting of an elf clear stone building, dark pine forest, magic the gathering, masterpiece, intricate details
Painting of an elf stone building in a dark forest, pine tree, magic the gathering, masterpiece, intricate details
Painting of a small troglodyte town on hills, rounded doors, bright green lands, small bushes, clear sky, masterpiece, intricate details
Isometric 3d rendering of a single medieval sheep barn on blank background
Isometric 3d rendering of a single medieval sheep barn building, masterpiece, intricate details
Isometric 3d rendering of a single sheep barn building, masterpiece, intricate details
Painting of a small troglodyte town, rounded doors, bright green lands, lush forest, clear sky, masterpiece, intricate details
Painting of a small troglodyte town, bright green lands, lush forest, clear sky, masterpiece, intricate details
Painting of a small hobbit town, bright green lands, lush forest, clear sky, masterpiece, intricate details
Painting of a small hobbit fantasy town, bright green lands, lush forest, clear sky, masterpiece, intricate details
DSLR photo of an albatros, soaring through the clouds above the ocean, amazing light, award winning
Painting of a small orc fantasy town, dark lands, gloomy, pine forest, fog, dark sky, masterpiece, intricate details
Painting of a small coastal elvish fantasy town, ocean on the left, fields on the right, fog, masterpiece, intricate details
DSLR photo of a small coastal fantasy town, mountains in the distance, ocean on the left, fields on the right, fog
DSLR wideshot photo of a small fantasy town in a valley, mountains in the background, crossed by river, forest on the left, fields on the right, fog
DSLR wideshot photo of a small fantasy orc town in a valley, mountains in the background, forest on the left, fields on the right, fog
DSLR wideshot photo of a small fantasy elvish town in a valley, mountains in the background, forest on the left, fields on the right, fog
DSLR wideshot photo of a small medieval town in a valley, mountains in the background, forest on the left, fields on the right, fog
DSLR wideshot photo of a small medieval town, mountain in the background, forest on the sides, fog
DSLR wideshot photo of a small medieval village, mountain in the background, forest on the sides, fog
DSLR wideshot photo of a country side medieval village, mountain in the background, forest of the side, clear weather
DSLR wideshot photo of a country side medieval village, snowy mountain in the background, forest of the side, clear weather
DSLR high angle photo of an amazing woman naked standing in the subway
DSLR high angle photo of an amazing woman naked lying on a bed


generator.py
import torch, os, re, random
from datetime import datetime
from torch import autocast
from diffusers import StableDiffusionPipeline, DPMSolverMultistepScheduler, EulerDiscreteScheduler
from diffusers.pipelines.stable_diffusion import safety_checker

class Generator:

    def set_scheduler_for_model(self):
        if self.model_id == "stabilityai/stable-diffusion-2-1":
            self.pipe = StableDiffusionPipeline.from_pretrained(self.model_id, torch_dtype=torch.float16)
            # self.pipe.scheduler = DPMSolverMultistepScheduler.from_config(self.pipe.scheduler.config) 
        elif self.model_id == "prompthero/openjourney":
            self.pipe = StableDiffusionPipeline.from_pretrained(self.model_id, torch_dtype=torch.float16)
            # self.pipe.scheduler = EulerDiscreteScheduler.from_pretrained(self.model_id, subfolder="scheduler")

    def __init__(self) -> None:
        if torch.cuda.is_available():
            print('Everything is okay')
        else:
            print('Shit fuck and bubblegum!')
            exit(255)
        self.device = "cuda"
        self.model_id = "stabilityai/stable-diffusion-2-1"
        self.model_id = "prompthero/openjourney"
        self.set_scheduler_for_model()
        self.pipe = self.pipe.to(self.device)
        safety_checker.StableDiffusionSafetyChecker.forward = self.sc

    def sc(self, clip_input, images) :
        return images, [False for i in images]


    def get_positive_negative_prompt(self, prompt: str) -> tuple:
        # Regular expression to match the desired pattern
        pattern = r'~(.*?),'

        # Extract the matching parts using regex
        matches = re.findall(pattern, prompt)

        # Concatenate the extracted parts into a new string
        negative_prompt = 'extra lib, too many fingers,' + ','.join(matches)

        # Remove the extracted parts from the original string
        positive_prompt = re.sub(pattern, '', prompt)
        return (positive_prompt, negative_prompt)


    def generate_single(self, prompt, negative_prompt, step, seed, dir_path, idx, width=512, heigth=512):
        today = datetime.today().strftime('%Y_%m_%d')
        if not os.path.exists(today):
            os.mkdir(today)
        dir_path = f"{today}\{dir_path}"
        with autocast(self.device):
            if self.model_id == 'prompthero/openjourney':
                prompt = f"mdjrny-v4 style, {prompt}"
            try:
                generator = torch.Generator("cuda").manual_seed(seed+idx)
                if not os.path.exists(dir_path):
                    os.mkdir(dir_path)
                image = self.pipe(prompt, negative_prompt=negative_prompt, guidance_scale=9, num_inference_steps=step, generator=generator, width=width, height=heigth).images[0]
                image_name = f'{dir_path}\\{idx}_{step}_{prompt.replace(" ", "_").replace(",", "").replace("|", "").replace(":", "").replace(".", "")[:120]}.png'
                if os.path.exists(image_name):
                    image_name = image_name[:-4] + '(2).png'
                image.save(image_name)
                return (image_name, seed+idx)
            except Exception as e:
                print(f'Exception: {e}')
                pass


    def generate(self, prompt, negative_prompt, step, seed, generate_count):
        # edit StableDiffusionSafetyChecker class so that, when called, it just returns the images and an array of True values

        # scheduler = EulerDiscreteScheduler.from_pretrained(model_id, subfolder="scheduler")
        # pipe = StableDiffusionPipeline.from_pretrained(model_id, revision="fp16", torch_dtype=torch.float16)

        with autocast(self.device):
            prompt = f"mdjrny-v4 style, {prompt}"
            try:
                dir_path = str(hash(prompt * random.randint(0, 10000)))
                for i in range(0, generate_count):
                    self.generate_single(prompt, negative_prompt, step, seed, dir_path, i)
            except Exception as e:
                print(f'Exception: {e}')
                pass


    # An hyperrealistic picture of Weeny the Poo, smoking a big joint, being stoned, surrounded by nature, sitted on the ground, its back against a tree

static/js/common.js
function createLoadingImage() {
    const existingElement = document.getElementById('loadingDiv');
    if (existingElement !== null) {
        return existingElement;
    }
    const loadingDiv = document.createElement('div');
    loadingDiv.id = 'loadingDiv';
    loadingDiv.classList = 'd-flex align-items-center justify-content-center';
    const spinner = document.createElement('div');
    spinner.classList = 'spinner-border text-primary';
    spinner.setAttribute('role', 'status');
    // const spanLoading = document.createElement('span');
    // spanLoading.classList = 'visually-hidden';
    // spanLoading.textContent = 'Loading...';
    // spinner.appendChild(spanLoading);
    loadingDiv.appendChild(spinner);
    return loadingDiv;
}

function downloadImage(id) {
    // Get the base64-encoded image data
    var imageBase64 = document.getElementById(id).getAttribute('src').replace(/^data:image\/(png|jpeg|jpg);base64,/, '');

    // Convert the base64-encoded data to a Blob object
    var blob = b64toBlob(imageBase64, 'image/jpeg');

    // Create a temporary <a> element and click it to initiate the download
    var link = document.createElement('a');
    link.download = id + '.jpg'; // Set the file name for the download
    link.href = window.URL.createObjectURL(blob);
    link.click();
}

function b64toBlob(b64Data, contentType, sliceSize) {
    contentType = contentType || '';
    sliceSize = sliceSize || 512;

    var byteCharacters = atob(b64Data);
    var byteArrays = [];

    for (var offset = 0; offset < byteCharacters.length; offset += sliceSize) {
        var slice = byteCharacters.slice(offset, offset + sliceSize);

        var byteNumbers = new Array(slice.length);
        for (var i = 0; i < slice.length; i++) {
            byteNumbers[i] = slice.charCodeAt(i);
        }

        var byteArray = new Uint8Array(byteNumbers);
        byteArrays.push(byteArray);
    }

    var blob = new Blob(byteArrays, { type: contentType });
    return blob;
}

function clearImages() {
    const imagesBlock = document.getElementById('images');
    while (imagesBlock.firstChild) {
        imagesBlock.removeChild(imagesBlock.firstChild);
    }
}

function randomString(length) {
    let result = '';
    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    const charactersLength = characters.length;

    for (let i = 0; i < length; i++) {
        result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }

    return result;
}

static/js/socket.js
var socket = io.connect(window.location.href);
var current_to_gen = 0;
socket.on('open', function (event) {
    console.log("WebSocket connected");
});
socket.on('close', function (event) {
    console.log("WebSocket closed");
});
socket.on('error', function (event) {
    console.log("WebSocket error: ${event}");
});
socket.on('message', function (message) {
    if (message.image != '') {
        displayImage(message.image, message.image_seed)
    }
    updateQueueLength(message.queue_length)
});

function sendMessage(prompt, negative, seed, step_count, generate_count, width, heigth) {
    var prompt = prompt || document.getElementById("prompt").value;
    var negative = negative || document.getElementById("negative").value;
    if (document.getElementById('basicNeg').checked) {
        negative += ', disfigured, kitschy, ugly, oversaturated, low-res, deformed, blurry, extra lib, too many fingers'
    }
    var seed = seed || document.getElementById("seed").value;
    var width = width || document.getElementById("width").value;
    var heigth = heigth || document.getElementById("heigth").value;
    var step_count = step_count || document.getElementById("step-count").value;
    var generate_count = generate_count || document.getElementById("generator-count").value;
    current_to_gen = generate_count;
    socket.emit("message", JSON.stringify(
        {
            "prompt": prompt,
            "negative": negative,
            "seed": seed,
            "step_count": step_count,
            "generate_count": generate_count,
            "width": width,
            "heigth": heigth,
        }
    ));
    displayLoading(createLoadingImage());
}

function displayLoading(loadingDiv) {
    // Add loading animation if needed
    if (loadingDiv != null && loadingDiv.parentNode != null) {
        loadingDiv.parentNode.removeChild(loadingDiv);
    }
    if (current_to_gen > 0) {
        document.getElementById('loading_space').appendChild(loadingDiv);
    }
}

function displayImage(message_image, message_seed) {
    message_seed = message_seed || '' + Math.floor(Math.random() * 999999999);
    var imageb64 = 'data:image/jpeg;base64,' + message_image;
    current_to_gen -= 1;

    const imagesBlock = document.getElementById('images');

    const imageDivTag = document.createElement('div');
    const imageLink = document.createElement('a');
    const imageTag = document.createElement('img');
    const imageSeedText = document.createElement('p');
    imageDivTag.classList = 'displayImage';
    imageLink.appendChild(imageTag)
    imageDivTag.appendChild(imageLink);
    imageDivTag.appendChild(imageSeedText);
    if (imagesBlock.firstChild) {
        imagesBlock.insertBefore(imageDivTag, imagesBlock.firstChild);
    } else {
        imagesBlock.appendChild(imageDivTag)
    }

    const id = message_seed + randomString(10)

    imageLink.onclick = () => downloadImage(id);
    imageLink.href = '#';
    imageTag.src = imageb64;
    imageTag.id = id
    imageTag.style = 'margin: 2px';
    imageSeedText.innerHTML = message_seed + '<a href="#"><i class="bi bi-arrow-counterclockwise" onclick="reprocess('+message_seed+')"></i></a>'
    imageSeedText.classList = 'displaySeed';

    displayLoading(createLoadingImage())
}

function updateQueueLength(l) {
    var divQueueInfoSpan = document.getElementById('queue_length_info_span');
    divQueueInfoSpan.textContent = "Messages in queue: " + l;
    current_to_gen = l
}

function clearGenQueue() {
    socket.emit("clear")
}

server.py
import os
import signal
import threading
import json
import random
import generator
import base64
from flask import Flask, render_template, send_from_directory
import flask_socketio as fs
from flask_socketio import SocketIO

gen = None
app = Flask(__name__)
socketio = SocketIO(app)
prompts = []

def signal_handler(sig, frame):
    print("Vous avez appuyé sur CTRL+C!")
    exit(0)

signal.signal(signal.SIGINT, signal_handler)

lock = threading.Lock()
queue_length = 0

def broadcast_queue_lenght():
    global queue_length
    fs.send({'queue_length': queue_length, 'image': ''}, broadcast=True)


def add_prompt_to_history(prompt):
    global prompts
    print(f'Prompts are: {prompts}')
    if prompt not in prompts:
        prompts.insert(0, prompt)
    with open('static/history', 'w') as history:
        for p in prompts:
            history.write(f'{p}\n')


def process_queue(message):
    global queue_length
    message = json.loads(message)
    prompt = message.get('prompt', '')
    add_prompt_to_history(prompt=prompt)
    negative = message.get('negative', '')
    steps = int(message.get('step_count', '60'))
    seed = message.get('seed', '')
    width = int(message.get('width', 512))
    heigth = int(message.get('heigth', 512))
    if seed == '':
        seed = random.randint(0, 99999999999999)
    else:
        seed = int(seed)
    to_generate = int(message.get('generate_count', 4))
    queue_length += to_generate
    broadcast_queue_lenght()
    dir_path = str(hash(prompt * random.randint(0, 10000)))
    with lock:
        print(f'Message: {json.dumps(message, indent=2)}')
        global gen
        if not gen:
            gen = generator.Generator()
        for idx in range(0, to_generate):
            if queue_length <= 0:
                queue_length = 0
                break
            (image_path, image_seed) = gen.generate_single(prompt=prompt, negative_prompt=negative, step=steps, seed=seed, dir_path=dir_path, idx=idx, width=width, heigth=heigth)
            with open(image_path, "rb") as f:
                image_data = f.read()
            image64 = base64.b64encode(image_data).decode('utf-8')
            queue_length -= 1
            fs.send({'image': image64, 'image_seed': image_seed, 'queue_length': queue_length})
            broadcast_queue_lenght()
    if queue_length <= 0:
        queue_length = 0

@app.route('/static/js/<path:path>')
def send_js(path):
    return send_from_directory('static/js', path)

@app.route('/static/css/<path:path>')
def send_css(path):
    return send_from_directory('static/css', path)

@app.route('/')
def index():
    return render_template('index.html')

@socketio.on('connect')
def handle_connect():
    broadcast_queue_lenght()
    print('Client connected')

@socketio.on('disconnect')
def handle_disconnect():
    print('Client disconnected')

@socketio.on('message')
def handle_message(message):
    broadcast_queue_lenght()
    process_queue(message=message)

@socketio.on('clear')
def handle_clear_queue():
    global queue_length
    queue_length = 0
    broadcast_queue_lenght()

if __name__ == '__main__':
    if not os.path.exists('static/history'):
        with open('static/history', 'w'):
            pass
    with open('static/history', 'r') as history:
        for p in history:
            if p.strip() != '':
                prompts.append(p.strip())
    socketio.run(app, host='0.0.0.0', port=8000, debug=True)